mavros_extras/src/plugins/gimbal_control.cpp

```diff
@@ -0,0 +1,690 @@
+      "~/manager/set_pitchyaw", 10, std::bind(
+        &GimbalControlPlugin::manager_set_pitchyaw_cb, this,
+        _1));
+
+    // --Not successfully validated-- also note that the message is the same as pitchyaw and will likely change
+    set_manager_manual_control_sub = node->create_subscription<mavros_msgs::msg::GimbalManagerSetPitchyaw>(
+      "~/manager/set_manual_control", 10, std::bind(
+        &GimbalControlPlugin::manager_set_manual_control_cb, this,
+        _1));
+
+
+    // Publishers
+    gimbal_attitude_status_pub = node->create_publisher<mavros_msgs::msg::GimbalDeviceAttitudeStatus>(
+      "~/device/attitude_status",
+      10);
+
+    gimbal_manager_status_pub = node->create_publisher<mavros_msgs::msg::GimbalManagerStatus>(
+      "~/manager/status", 
+      10);
+
+    gimbal_manager_info_pub = node->create_publisher<mavros_msgs::msg::GimbalManagerInformation>(
+      "~/manager/info", 
+      10);
+
+    // --Not successfully validated--
+    gimbal_device_info_pub = node->create_publisher<mavros_msgs::msg::GimbalDeviceInformation>(
+      "~/device/info", 
+      10);
+
+
+    // Services
+    // --Not successfully validated--
+    gimbal_device_info_srv = node->create_service<mavros_msgs::srv::GimbalGetInformation>(
+      "~/device/get_info", std::bind(
+        &GimbalControlPlugin::device_get_info_cb,
+        this, _1, _2));
+    
+    gimbal_manager_info_srv = node->create_service<mavros_msgs::srv::GimbalGetInformation>(
+      "~/manager/get_info", std::bind(
+        &GimbalControlPlugin::manager_get_info_cb,
+        this, _1, _2));
+
+    gimbal_manager_configure_srv = node->create_service<mavros_msgs::srv::GimbalManagerConfigure>(
+      "~/manager/configure", std::bind(
+        &GimbalControlPlugin::manager_configure_cb,
+        this, _1, _2));
+
+    gimbal_manager_pitchyaw_srv = node->create_service<mavros_msgs::srv::GimbalManagerPitchyaw>(
+      "~/manager/pitchyaw", std::bind(
+        &GimbalControlPlugin::manager_pitchyaw_cb,
+        this, _1, _2));
+
+    gimbal_manager_set_roi_srv = node->create_service<mavros_msgs::srv::GimbalManagerSetRoi>(
+      "~/manager/set_roi", std::bind(
+        &GimbalControlPlugin::manager_set_roi_cb,
+        this, _1, _2));
+
+    // --Not successfully validated--
+    gimbal_manager_camera_track = node->create_service<mavros_msgs::srv::GimbalManagerCameraTrack>(
+      "~/manager/camera_track", std::bind(
+        &GimbalControlPlugin::manager_camera_track,
+        this, _1, _2));
+    
+    // Client used by all services for sending mavros/cmd/command service calls, on a separate callback group to support nested service calls
+    cmdClient = node->create_client<mavros_msgs::srv::CommandLong>("cmd/command", rmw_qos_profile_services_default, cb_group);
+    while (!cmdClient->wait_for_service(std::chrono::seconds(5))) {
+        RCLCPP_ERROR(node->get_logger(), "GimbalControl: mavros/cmd/command service not available after waiting");
+    }
```

Comment:
Original comment by vooon:
> Locking in the constructor might be a bad idea.
> Maybe better to check for that service later?

As the rest is callbacks based on ros/mavlink message traffic, then you would have to add a check to every callback needing this client, instead of having it in the constructor once. 
The locking here is also set with a timeout of 5s, so worst-case is locked for 5s before moving on. 

If you have a suggestion to how to handle it better, let me know :)

Reply:
edit: I miswrote. I can see it will block here in periods of 5s forever. 
Potential change could be to make it to only try once with that timeout and then abandon the plugin and throw the error, if the service can't be made?

Reply:
I can change to do that, I just thought the constant recreation of the client would be worse? 
As if a client is using the service to set ROI region, or pitchyaw with even just 1hz it is still initializing a client and destroying it per call. 

I could create it in the constructor, but move the check if the service is available to each message using it? And change so it only tries for 5s then just returns error to the caller to let them know? 

Would keep the constructor free from blocking, but still ensure the service exist before calling it, and otherwise return message to caller that the request could not be completed due to the mavros/cmd service not being found?

Reply:
You can cache the client, but must create it on request, not in the constructor.

Something like that pseudo:
```cpp
*Client get_cmd_long() {
  mutex Locker();

  if _cache {
    return _cache;
  }

  client = node_->create_client(...);
  // error handling!
  _cache = client;
  return client;  
}
```

Reply:
ofc, that makes sense. I'll do that.